File: runtimes/ruby/versions/latest/src/logger.rb
require 'securerandom'

class RuntimeLogger
  TYPE_LOG = "log"
  TYPE_ERROR = "error"

  attr_accessor :enabled
  attr_accessor :id
  attr_accessor :includes_native_info
  attr_accessor :stream_logs
  attr_accessor :stream_errors
  attr_accessor :custom_std_stream
  attr_accessor :native_logs_cache
  attr_accessor :native_errors_cache

  def initialize(status, id)
    @includes_native_info = false

    @enabled = false

    if status.nil? || status.empty? || status === "enabled"
      @enabled = true
    end

    if @enabled === true
      serverEnv = ENV['OPEN_RUNTIMES_ENV'] || ""

      if id.nil? || id.empty?
        if serverEnv === "development"
          @id = "dev"
        else
          @id = self.generate_id
        end
      else
        @id = id
      end

      @stream_logs = File.open("/mnt/logs/" + @id + "_logs.log", 'a')
      @stream_errors = File.open("/mnt/logs/" + @id + "_errors.log", 'a')
    end
  end

  def write(messages, type = RuntimeLogger::TYPE_LOG, native = false)
    if !@enabled
      return
    end

    if native && !@includes_native_info
      @includes_native_info = true
      self.write(["Native logs detected. Use context.log() or context.error() for better experience."], type,
                 native)
    end

    stream = @stream_logs

    if type === RuntimeLogger::TYPE_ERROR
      stream = @stream_errors
    end

    string_log = ""
    i = 0
    messages.each do |message|
      if message.kind_of?(Array) || message.kind_of?(Hash)
        string_log += message.to_json
      else
        string_log += message.to_s
      end

      if i < messages.length() - 1
        string_log += " ";
      end

      i += 1
    end

    try {
      stream.write(string_log)
    } catch (Exception e) {
      # Silently fail to prevent 500 errors in runtime
      # Log write failures should not crash the runtime
    }
  end

  def end()
    if !@enabled
      return
    end

    @enabled = false

    @stream_logs.close
    @stream_errors.close
  end

  def override_native_logs()
    @native_logs_cache = $stdout
    @native_errors_cache = $stderr
    @custom_std_stream = StringIO.new
    $stdout = @custom_std_stream
    $stderr = @custom_std_stream
  end

  def revert_native_logs()
    $stdout = @native_logs_cache
    $stderr = @native_errors_cache

    unless @custom_std_stream.string.nil? || @custom_std_stream.string.empty?
      self.write([custom_std_stream.string], RuntimeLogger::TYPE_LOG, true)
    end
  end

  def generate_id(padding = 7)
    now = Time.now
    sec = now.to_i
    usec = now.usec
    base_id = "%08x%05x" % [sec, usec]
    random_padding = SecureRandom.hex(padding)
    random_padding = random_padding[0...padding]
    base_id + random_padding
  end
end

File: runtimes/python/versions/latest/src/logger.py
import json
import sys
from datetime import datetime
import math
import os
from io import StringIO


class Logger:
    TYPE_ERROR = "error"
    TYPE_LOG = "log"

    id = ""
    enabled = False
    includes_native_info = False

    stream_logs = None
    stream_errors = None

    custom_std = None

    def __init__(self, status=None, id=None):
        if status is None or status == "":
            self.enabled = True
        else:
            if status == "enabled":
                self.enabled = True
            else:
                self.enabled = False

        if self.enabled is True:
            is_development = False
            if os.getenv("OPEN_RUNTIMES_ENV") == "development":
                is_development = True

            if id is None or id == "":
                if is_development is True:
                    self.id = "dev"
                else:
                    self.id = self.generate_id()
            else:
                self.id = id

            self.stream_logs = open("/mnt/logs/" + self.id + "_logs.log", "a")
            self.stream_errors = open("/mnt/logs/" + self.id + "_errors.log", "a")

    def write(self, messages, xtype=None, is_native=False):
        if xtype is None:
            xtype = Logger.TYPE_LOG

        if self.enabled is False:
            return

        if is_native is True and self.includes_native_info is False:
            self.includes_native_info = True
            self.write(
                [
                    "Native logs detected. Use context.log() or context.error() for better experience."
                ]
            )

        stream = self.stream_logs

        if xtype == Logger.TYPE_ERROR:
            stream = self.stream_errors

        string_log = ""
        i = 0
        for message in messages:
            if isinstance(message, (list, dict, tuple)):
                string_log += json.dumps(message, separators=(",", ":"))
            else:
                string_log += str(message)

            if i < len(messages) - 1:
                string_log += " "

            i += 1

        try:
            stream.write(string_log)
        except Exception as e:
            # Silently fail to prevent 500 errors in runtime
            # Log write failures should not crash the runtime
            pass

    def end(self):
        if self.enabled is False:
            return

        self.enabled = False

        self.stream_logs.close()
        self.stream_errors.close()

    def override_native_logs(self):
        sys.stderr = sys.stdout = self.custom_std = StringIO()

    def revert_native_logs(self):
        if self.custom_std is not None and self.custom_std.getvalue():
            self.write([self.custom_std.getvalue()], Logger.TYPE_LOG, True)

        sys.stdout = sys.__stdout__
        sys.stderr = sys.__stderr__

    def generate_id(self, padding=7):
        now = datetime.now()
        sec = int(now.timestamp())
        usec = now.microsecond % 1000
        base_id = f"{sec:08x}{usec:05x}"
        random_bytes = os.urandom(math.ceil(padding / 2))
        random_padding = random_bytes.hex()[:padding]
        return base_id + random_padding
